Inductor cpp wrapper: fix codegen of FallbackKernel with kwargs

chunyuan-w opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104575

Fix cpp wrapper failure on TorchBench model `hf_Reformer` with `randn`:
```
random_rotations = torch.randn(rotations_shape, device=vectors.device, dtype=vectors.dtype)
```

For cpp wrapper, when `kwargs` is not empty, for `OpOverloadPacket` kernel, we need to know the exact overload schema to handle the `kwargs` properly when calling the cpp kernel: including finding the correct order of the kwargs and getting the default value for optional args without provided value when calling the function (`layout` in the above case).

The current support in this PR is conservative and we'll extend the functionality in subsequent PRs.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

chunyuan-w committed:
Inductor cpp wrapper: fix codegen of FallbackKernel with kwargs

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix codegen of FallbackKernel with kwargs"


Fix cpp wrapper failure on TorchBench model `hf_Reformer` with `randn`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix codegen of FallbackKernel with kwargs"


Fix cpp wrapper failure on TorchBench model `hf_Reformer` with `randn`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix codegen of FallbackKernel with kwargs"


Fix cpp wrapper failure on TorchBench model `hf_Reformer` with `randn`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix codegen of FallbackKernel with kwargs"


Fix cpp wrapper failure on TorchBench model `hf_Reformer` with `randn`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix codegen of FallbackKernel with kwargs"


Fix cpp wrapper failure on TorchBench model `hf_Reformer` with `randn`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix codegen of FallbackKernel with kwargs"


Fix cpp wrapper failure on TorchBench model `hf_Reformer` with `randn`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix codegen of FallbackKernel with kwargs"


Fix cpp wrapper failure on TorchBench model `hf_Reformer` with `randn`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix codegen of FallbackKernel with kwargs"


Fix cpp wrapper failure on TorchBench model `hf_Reformer` with `randn`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix codegen of FallbackKernel with kwargs"


Fix cpp wrapper failure on TorchBench model `hf_Reformer` with `randn`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix codegen of FallbackKernel with kwargs"


Fix cpp wrapper failure on TorchBench model `hf_Reformer` with `randn`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1173,7 +1174,7 @@ def generate_profiler_mark_wrapper_call(self, stack):
         )
 
-    def codegen_device(self, device):
+    def codegen_option(self, device):
```
Can we just fix `codegen_device` without adding the new `codegen_option`?

chunyuan-w replied:
Refactored the code to reuse `codegen_device` in `codegen_option`.
I checked that both of them are needed.

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix codegen of FallbackKernel with kwargs"


Fix cpp wrapper failure on TorchBench model `hf_Reformer` with `randn`:
```
random_rotations = torch.randn(rotations_shape, device=vectors.device, dtype=vectors.dtype)
```

For cpp wrapper, when `kwargs` is not empty, for `OpOverloadPacket` kernel, we need to know the exact overload schema to handle the `kwargs` properly when calling the cpp kernel: including finding the correct order of the kwargs and getting the default value for optional args without provided value when calling the function (`layout` in the above case).

The current support in this PR is conservative and we'll extend the functionality in subsequent PRs.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix codegen of FallbackKernel with kwargs"


Fix cpp wrapper failure on TorchBench model `hf_Reformer` with `randn`:
```
random_rotations = torch.randn(rotations_shape, device=vectors.device, dtype=vectors.dtype)
```

For cpp wrapper, when `kwargs` is not empty, for `OpOverloadPacket` kernel, we need to know the exact overload schema to handle the `kwargs` properly when calling the cpp kernel: including finding the correct order of the kwargs and getting the default value for optional args without provided value when calling the function (`layout` in the above case).

The current support in this PR is conservative and we'll extend the functionality in subsequent PRs.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

jgong5 commented on torch/_inductor/utils.py:
```diff
@@ -1198,13 +1198,6 @@ def is_optional(arg):
     return "Optional" in str(arg.type)
```
Make this function inner too?

chunyuan-w replied:
Fixed

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix codegen of FallbackKernel with kwargs"


Fix cpp wrapper failure on TorchBench model `hf_Reformer` with `randn`:
```
random_rotations = torch.randn(rotations_shape, device=vectors.device, dtype=vectors.dtype)
```

For cpp wrapper, when `kwargs` is not empty, for `OpOverloadPacket` kernel, we need to know the exact overload schema to handle the `kwargs` properly when calling the cpp kernel: including finding the correct order of the kwargs and getting the default value for optional args without provided value when calling the function (`layout` in the above case).

The current support in this PR is conservative and we'll extend the functionality in subsequent PRs.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix codegen of FallbackKernel with kwargs"


Fix cpp wrapper failure on TorchBench model `hf_Reformer` with `randn`:
```
random_rotations = torch.randn(rotations_shape, device=vectors.device, dtype=vectors.dtype)
```

For cpp wrapper, when `kwargs` is not empty, for `OpOverloadPacket` kernel, we need to know the exact overload schema to handle the `kwargs` properly when calling the cpp kernel: including finding the correct order of the kwargs and getting the default value for optional args without provided value when calling the function (`layout` in the above case).

The current support in this PR is conservative and we'll extend the functionality in subsequent PRs.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

chunyuan-w commented on torch/_inductor/utils.py:
```diff
@@ -1156,3 +1156,107 @@ def red_text(msg):
+    if min_args != max_pos_args:
+        return f"takes from {min_args} to {max_pos_args} positional arguments but {nargs} were given"
+    else:
+        return f"takes {max_pos_args} positional arguments but {nargs} were given"
```
Fixed

chunyuan-w commented:
@pytorchbot rebase

chunyuan-w committed:
Update on "Inductor cpp wrapper: fix codegen of FallbackKernel with kwargs"


Fix cpp wrapper failure on TorchBench model `hf_Reformer` with `randn`:
```
random_rotations = torch.randn(rotations_shape, device=vectors.device, dtype=vectors.dtype)
```

For cpp wrapper, when `kwargs` is not empty, for `OpOverloadPacket` kernel, we need to know the exact overload schema to handle the `kwargs` properly when calling the cpp kernel: including finding the correct order of the kwargs and getting the default value for optional args without provided value when calling the function (`layout` in the above case).

The current support in this PR is conservative and we'll extend the functionality in subsequent PRs.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

chunyuan-w commented:
@pytorchbot merge

chunyuan-w commented:
@pytorchbot merge
